src/Platforms/Echo.Platforms.AsmResolver/Emulation/Invocation/DelegateInvoker.cs

```diff
@@ -0,0 +1,100 @@
+            return InvocationResult.Inconclusive();
+        
+        if (method.Name == ".ctor")
+        {
+            return ConstructDelegate(context, arguments);
+        }
+        
+        if (method.Name == "Invoke")
+        {
+            return InvokeDelegate(context, arguments);
+        }
+
+        return InvocationResult.Inconclusive();
+    }
+    
+    private InvocationResult ConstructDelegate(CilExecutionContext context, IList<BitVector> arguments)
+    {
+        var vm = context.Machine;
+        var corlib = vm.ContextModule.CorLibTypeFactory;
+
+        var self = arguments[0].AsObjectHandle(vm);
+        var obj = arguments[1];
+        var methodPtr = arguments[2];
+
+        var _delegate = self.GetObjectType().Resolve()!.BaseType!.Resolve()!.BaseType;
+        IFieldDescriptor _target = new MemberReference(_delegate, "_target", new FieldSignature(corlib.Object));
+        IFieldDescriptor _methodPtr = new MemberReference(_delegate, "_methodPtr", new FieldSignature(corlib.IntPtr));
+
+        if (obj.IsFullyKnown && obj.AsObjectHandle(vm).IsNull)
+            self.WriteField(_target, obj);
+
+        self.WriteField(_methodPtr, methodPtr);
+
+        //var methodBase = vm.FunctionMarshaller.ResolveMethodPointer((nuint)methodPtr.AsSpan().ReadNativeInteger(vm.Is32Bit));
+        //if (methodBase != null)
+        //    self.WriteField(_methodBase, marshall IMethodDescriptor to RuntimeMethodInfo);
+
+        return InvocationResult.StepOver(null);
+    }
+    
+    private InvocationResult InvokeDelegate(CilExecutionContext context, IList<BitVector> arguments)
+    {
+        var vm = context.Machine;
+        var stack = context.CurrentFrame.EvaluationStack;
+        var corlib = vm.ContextModule.CorLibTypeFactory;
+
+        var self = arguments[0].AsObjectHandle(vm);
+
+        var _delegate = self.GetObjectType().Resolve()!.BaseType!.Resolve()!.BaseType;
+        IFieldDescriptor _target = new MemberReference(_delegate, "_target", new FieldSignature(corlib.Object));
+        IFieldDescriptor _methodPtr = new MemberReference(_delegate, "_methodPtr", new FieldSignature(corlib.IntPtr));
+
+        var methodPtr = self.ReadField(_methodPtr).AsSpan().ReadNativeInteger(vm.Is32Bit);
+        var method = vm.FunctionMarshaller.ResolveMethodPointer((nuint)methodPtr);
+
+        if (method == null)
+            throw new Exception("Cant resolve method");
+
+        var arg1 = self.ReadField(_target);
+        if (arg1.AsSpan().ReadNativeInteger(vm.Is32Bit) != 0)
+            stack.Push(arg1.AsObjectHandle(vm));
+
+        // skip 1 for delegate "this"
+        for (var i = 1; i < arguments.Count; i++)
+            stack.Push(arguments[i], method.Signature!.ParameterTypes[i]);
+
+        var result = vm.Dispatcher.Dispatch(context, new(CilOpCodes.Call, method));
+
+        if (!result.IsSuccess)
+            return InvocationResult.Exception(result.ExceptionObject);
+
+        return InvocationResult.StepOver(null);
```

Comment:
I feel this is problematic or at the very least inaccurate:

Firstly, this goes through CIL instruction dispatching when in real processes this does not happen. I feel we should feed the invocation back to the VM's method invoker and return its invocation result. This also avoids a lot of unnecessary marshalling to stack slots and back.

Additionally, skipping the call with `StepOver(null)` results in the `Invoke` method call actually not being registered as a call in the callstack of the thread. This isn't usually a problem for other shims (such as methods in `Unsafe`) as we intend to immediately return from these methods anyway. However, depending on the VM's method invoker, we may actually need to step into the method referenced by the delegate. Not having `Invoke` in the callstack then is not just inaccurate emulation, but can also be detected by the callee when we get to actually emulating call stacks in `StackTrace` and `Exception` objects.

Reply:
I have two ideas:
1) A FakeCallFrame that will only show that this method was called. (lots of virtual and overload ctrl+c ctrl+v)
2) Create (if MethodDefinition.CilMethodBody == null) a CilMethodBody for Invoke method, which will be like: optional(ldarg_0 -> ldfld _target) -> ldarg [0..N] -> ldarg_0 -> ldfld _methodptr -> calli -> ret. And return StepIn or how it works.

Reply:
Hmm this is a tough choice. Neither option is great, and seems to highlight a fundamental limitation in the current invoker API. It makes me wonder if `InvocationResult` should get another result type, something like `InvocationResultType::FullyHandled` to allow method invokers to fully override the call behavior for extreme cases like these.

With such a result type, it then may make the most sense to let the delegate invoker push a normal `CallFrame` for the `Invoke` method, and then off-load to the VM's method invoker.
- If that results in a step-in, we keep the `Invoke` call frame on the stack before returning `FullyHandled`.
- For any other result, we can pop the `Invoke` call frame again before returning said result.

The only open question then is how to handle the final return from the `Invoke` method to the caller.

Reply:
Something like that? [commit](https://github.com/Washi1337/Echo/pull/140/commits/b2aebc0c2c8370c5615b403398e79947ad79e9e4)

Reply:
I like the idea of `IsTrampoline`, that works out nicely.

However like I said before, I really think `DelegateInvoker` should not be the one setting up the underlying call itself and instead use `context.Machine.Invoker` to do this. This way users can still customize (i.e. shim) calls that are behind delegates.

Reply:
I forgot about Invoker. I made a new commit.